### PR TITLE
Update compiler-rt to LLVM 17.0.4

### PR DESCRIPTION
**Pull request by aheejin**
On top of the main library code changes, each fix that was necessary for Emscripten was committed separately with its own commit message.

### CODE REVIEW

**Comment by aheejin** [comment]
Ping 😀

File: system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp [hunk 1]
```diff
@@ -453,17 +453,32 @@ INTERCEPTOR(int, pthread_create, void *th, void *attr,
   ENSURE_LSAN_INITED;
   EnsureMainThreadIDIsCorrect();
 
+#if !SANITIZER_EMSCRIPTEN
```
Comment (sbc100) [inline system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp:456]: I wonder if could just have a single `if` at the top of this function?  

```
#if !SANITIZER_EMSCRIPTEN
// Treat __ATTRP_C11_THREAD like the null attr
if (attr == __ATTRP_C11_THREAD) attr = nullptr;
#endif
```

**Review by sbc100 — APPROVED** [review]
lgtm % comment

File: system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp (hunk 1 above)
Comment (aheejin) [reply system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp:456]: That's a good idea. Btw you meant `#if SANITIZER_EMSCRIPTEN` (without `!`), right?

Comment (sbc100) [reply system/lib/compiler-rt/lib/lsan/lsan_interceptors.cpp:456]: Yup

**Comment by sbc100** [comment]
Oh, I forgot to say we should add something to the ChangeLog for this (and the libc++ change).

**Comment by aheejin** [comment]
> Oh, I forgot to say we should add something to the ChangeLog for this (and the libc++ change).

Done in #20736

**Pull request merged**